### DISCUSSION TITLE
Completing the Bluetooth LE port to the PICO-W

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#13612] [RP2040 / Pico]
I have posted discussion points related to this topic in the libraries category. But I have come to understand that many of the problems I was having were not due to the Bluetooth libraries but the PICO-W port. My understanding is that the C-SDK for PICO-W supports all the features of pairing/bonding and Resolvable Privacy Addressing (RPA). Both the MicroPython  libraries and the C-SDK rely on btstack so support SHOULD be possible.

**So the reason that aioble pairing and RPA does not work on the PICO-W is that the port is incomplete (I would guess).** The only generic pairing behavior missing in the aioble library is handling of passkeys and yes/no confirmations that could occur.

I would like to fix this but I could certainly use some help just to get started. This project is complicated and though I have considerable experience with using C on BLE embedded platforms (Nordic and Renesas) and Java on Android, my Python experience is limited. I tried playing around with some of the defines in the h-files and I got the client side to initiate a pairing request (whereas before it did not) but it did not continue past the peripheral response. I was really just guessing as I do not understand how these ports are set up.

Beyond completing the port there are also some documentation questions that would be helpful so one would know the following:

1. In pairing/bonding, who is responsible for saving the Long Term Keys (LTKs) and IRKs? Is it my application or is this done by the library?
2. For a bonded device, who is responsible for saving the results of service discovery and the enabled states of characteristics (setting them to indicate or notify or both), such that a peripheral device in a bonded reconnect can send data directly after encryption (without redoing pairing, service discovery, and characteristic enabling)? Is it my application or the library?
3. Is the 0x1800 and 0x1801 services and characteristics on the client side done by the library or does my application need to do it?
4. It appears that the application client is responsible for invoking pairing on an insufficient authentication error but the library appears to do it on the reception of a security request. This introduces a difficulty if my application is responsible for saving LTKs and IRKs. The same could be said for the insufficient encryption error but since I have been unable to pair I don't know what the library does.

If anyone has answers to these questions that would be helpful and it could be added to the documentation,

Why is this so important? Most medical devices require bonding and remote patient monitoring becomes impossible without it.

**Comment by andrewleech** [comment]
I'm just going by memory here which is from a few years ago really, I'll try to update this with more accurate details if I get time to check.

> 1. In pairing/bonding, who is responsible for saving the Long Term Keys (LTKs) and IRKs? Is it my application or is this done by the library?

`aioble.secrets` module manages these keys automatically, savings and loading then from a json file. The name and location of this file can be overridden if desired.

> 2. For a bonded device, who is responsible for saving the results of service discovery and the enabled states of characteristics 

I presume you mean for an application running as central? I haven't really used this much myself, I don't believe there is any built in saving of these details currently


> 3. Is the 0x1800 and 0x1801 services and characteristics on the client side done by the library or does my application need to do it?

The C level modbluetooth module sets up these services which are currently hardcoded. I did a bit of work updating them for "service changed" and database hash  support (unfinished) but don't remember how much of this gets exposed to the python code.


> 4. It appears that the application client is responsible for invoking pairing on an insufficient authentication error but the library appears to do it on the reception of a security request. 

By insufficient authentication request, do you mean when characteristics are flagged to require encryption (pairing)?
This is supposed to be handled automatically and does for me with a micropython ble peripheral (or did on stm at least). I'm not certain about central side though

**Comment by brianreinhold** [comment]
@andrewleech Thanks for your response. As far as saving service discovery is concerned that is something done by a bonded client. However, CCCDs are save by both the bonded client and peripheral.

Insufficient authentication is an error message sent by the peripheral when the client tries to read/write a secured characteristic/descriptor. In my aioble client implementation on the PICO-W (using btstack under the hood), the insufficient authentication error is passed up to the application as a GattError (contains just the number 5) and my application has to invoke pairing.

I should mention that there is no handler implemented in the Micropython distro built for PICO-W for pairing, so you have to write one and register it (done in modbluetooth_btstack.c). One can borrow code from the btstack examples which have such handlers and registrations. Without this added code, pairing does not work at all. However, we are still having problems getting pairing to work when the application calls the pair method from aioble (we need to do this when getting an insufficient authentication). It is possible more code is missing.

**Reply by andrewleech** [reply]
Ah I'm pretty sure I started work to store cccd's but not sure now where it ended up...

**Comment by brianreinhold** [comment]
@andrewleech At the moment we are trying to write a generic central for health devices on a PICO-W. We have found that there was no handler for security events in the `modbluetooth_btstack.c` file and have added one (taken from a `btstack` example: `sm_packet_handler` ) and registered this method in the `modbluetooth_btstack.c`. With that addition, we have been able to successfully bond and reconnect with health devices that send a `security request`. However, for those devices that do not send a `security request` what happens is that we will eventually get an `insufficient authentication` error when we try to write or read some characteristic/descriptor. It will always happen before we can get measurements. When this error is received, we have to handle pairing and therefore call the pairing method from `aioble`. It never succeeds. We get as far as the  `SM_EVENT_JUST_WORKS_REQUEST` or `SM_EVENT_PASSKEY_INPUT_NUMBER` event depending upon the device. We know that the passkey handling is incomplete but the 'just works' case should finish. Taking over-the-air sniffs shows that it is the central side that stops responding in the sm protocol sequence.

As far as remembering keys, that is definitely the case if we do not power cycle the PICO-W.
As far as saving service discovery and CCCDs, we have not done enough testing yet to see what happens here. Many health devices are poorly implemented and have errors in their protocol handling so we re-write the CCCDs every connection anyways. That is not very costly. But doing service discovery every time is.

It would be nice if the behavior were consistent with respect to pairing/bonding. Either the application handles `security requests` and `insufficient authentication`, or the library does. Now it is a mix.

If we could figure out how to get the logging in `btstack` to work it would be a great asset. Right now we have no idea how far the pairing actually gets before it fails.

**Reply by andrewleech** [reply]
That's great progress so far!
Have you enabled the debug printf at the top of modbluetooth_btstack ? 
I've had nimble stack logging going before, haven't really worked with btstack though sorry.

**Reply by brianreinhold** [reply]
Yes I have. That allows us to get the logging from the `modbluetooth_btstack.c` file but not the logging in the `btstack` itself.

**Comment by andrewleech** [comment]
Another suggestion that might help, there are "multitests" for bluetooth support in micropython itself which can be helpful for testing support at all the lower levels (before aioble gets involved)

These multitests instrument two devices at the same time to initiate connections and test results etc.

eg.
```
cd micropython/tests
python3 run-multitests.py -i pyb:/dev/ttyACM0 -i pyb:/dev/ttyACM1 ./multi_bluetooth/ble_gap_pair_bond.py

./multi_bluetooth/ble_gap_pair_bond.py on ttyACM0|ttyACM1: pass
1 tests performed
1 tests passed
```

That was run with a pyboard-d and a nrf52 board plugged in (running my MR to add nimble support there) to run the test.

If you look at the `./multi_bluetooth/ble_gap_pair_bond.py` you can see the functions that are run on each of the two boards, and then `./multi_bluetooth/ble_gap_pair_bond.py.exp` is the expected outputs the test outputs are compared against. So if you're printing anything else (eg with debug printouts enabled) it'll fail the test, but this can be helpful regardless.

```
anl@STEP: ~/micropython/tests $ ./run-multitests.py --help
usage: run-multitests.py [-h] [-s] [-t] [-i INSTANCE] [-p PERMUTATIONS] files [files ...]

Run network tests for MicroPython

positional arguments:
  files                 input test files

optional arguments:
  -h, --help            show this help message and exit
  -s, --show-output     show test output after running
  -t, --trace-output    trace test output while running
  -i INSTANCE, --instance INSTANCE
                        instance(s) to run the tests on
  -p PERMUTATIONS, --permutations PERMUTATIONS
                        repeat the test with this many permutations of the instance order

Supported instance types:
 -i pyb:<port>   physical device (eg. pyboard) on provided repl port.
 -i micropython  unix micropython instance, path customised with MICROPY_MICROPYTHON env.
 -i cpython      desktop python3 instance, path customised with MICROPY_CPYTHON3 env.
 -i exec:<path>  custom program run on provided path.
Each instance arg can optionally have custom env provided, eg. <cmd>,ENV=VAR,ENV=VAR...
```

```
anl@STEP: ~/micropython/tests $ ./run-multitests.py -t -i pyb:/dev/ttyACM0 -i pyb:/dev/ttyACM1 ./multi_bluetooth/ble_gap_pair_bond.py
./multi_bluetooth/ble_gap_pair_bond.py on ttyACM0|ttyACM1:
TRACE ttyACM0|ttyACM1:
   486 i0 : SET BDADDR = (0, b'\x02"2\x07"\xae')
   486 i0 : gap_advertise
   486 i0 : NEXT
   863 i1 : NEXT
   864 i0 : _IRQ_CENTRAL_CONNECT
   864 i1 : gap_connect
   864 i1 : _IRQ_PERIPHERAL_CONNECT
  1466 i1 : _IRQ_GATTC_CHARACTERISTIC_RESULT UUID('00000000-1111-2222-3333-444444444444')
  1466 i1 : _IRQ_GATTC_CHARACTERISTIC_DONE
  1466 i1 : gap_pair
  3071 i1 : _IRQ_ENCRYPTION_UPDATE 1 0 1
  3071 i1 : gattc_read
  3172 i0 : _IRQ_ENCRYPTION_UPDATE 1 0 1
  3172 i1 : _IRQ_GATTC_READ_RESULT b'encrypted'
  3172 i0 : _IRQ_GATTS_READ_REQUEST
  3172 i1 : gap_disconnect: True
  3273 i1 : _IRQ_PERIPHERAL_DISCONNECT
  3474 i0 : _IRQ_CENTRAL_DISCONNECT
pass
1 tests performed
1 tests passed
```

**Reply by brianreinhold** [reply]
@andrewleech we will be looking at all the tools we can. Thanks for the idea. We have managed to get logging from the btstack and that has proven crucial we can see that pairing initiated from aioble actually succeeds at the btstack level, it's the communication back up to python that is broken (somewhere). The python await times out and the device disconnects, but the C-code reports pairing success.

We would also like to have symmetry. Either the app does the pairing (for both security requests and insufficient authentication) or the btstack does it. Right now it's an unhealthy mixture of both. I would prefer the app to it to keep the exchanges synchronous. Synchronous behavior is the answer for robustness in the field. The app would start pairing and THEN invoke service discovery - no mix. Nice an clean,

**Reply by mradityagoyal** [reply]
Was this resolved? I can't seem to find a way to use a pico W as a Bluetooth central. I want to use off the shelf Bluetooth remotes (tictok scroller types) to send signals to my pico W... And I can't seem to find a Bluetooth pairing example using aioble.

**Reply by brianreinhold** [reply]
No it has not - at least not as a general solution. We have made modifications in modbluetooth_btstack.c (mainly handle all the pairing related events) and related files as well as modifications to aioble to handle pairing of health devices. I don't think we made any changes in the sm.c file of the btstack. There are still behaviors which pose difficulties mainly
1. The btstack handles security requests and responds by pairing or encrypting (bonded reconnect) and aioble and the application know nothing about it
2. The btstack does NOT handle insufficient authentication/encryption by pairing or encrypting but leaves that up to the application.
3. The btstack should either handle all of it or none of it. It's the two different ways to handle these events that cause a problem.

To solve this problem I had to add a new method to aioble's security module that just registered for encryption changes WITHOUT doing a pair request. Then I had to capture more events from sm.c and pass them up to the encryption changed in aioble, then in my application I had to run a task at connection which listened for encryption changed results using the new method in aioble. I also had to invoke a pairing from the app if there was insufficient authentication. To coordinate between the two I had to kill one if the other was needed. I did this with flags in aioble.

Service Discovery results are not saved and you have to do service discovery every connection. Many bonded health devices do not wait for service discovery to complete but once encrypted send their data and disconnect. This is a problem we have not been able to workaround yet.

Some health device issue security requests, others will issue insufficient authentications on unsecured reads/writes, and some require no pairing at all.

It's a complicated set of issues that need to be resolved to get pairing and bonded reconnects to work and I can't describe all of them here. I certainly do not have the expertise to make this part of the Micropython build.

I do not know your use case (I dont know what tictok scroller types are). Life will be much more straight forward if your peers always invoke security requests or never invoke security requests and always require pairing. You will still need to add a handler for btstack security methods in modbluetooth_btstack.c.

**Reply by mradityagoyal** [reply]
Thanks for the info!!

https://a.co/d/7xGJytk is what I am trying to pair with. (Pico W as central and that gamepad as a peripheral). 

Could you share the handler method code you added? 

Luckily, my device stack is less messy than the healthcare use case you are working on.

**Reply by mradityagoyal** [reply]
Also.. have you tried the esp32 nimble stack? 

I am going to give that a try as it seems pairing works on that stack...
https://docs.micropython.org/en/latest/library/bluetooth.html#pairing-and-bonding

**Reply by brianreinhold** [reply]
I have not. In our case requirements were such that we had to use the PICO_W. However, I do realize that other platforms are fully supported; it was just not an option for us.
